### ISSUE TITLE
Why was this abandoned?

### ISSUE THREAD

**Issue by lf94** [WebReflection/node-gtk#36]
Simple as that. I've been working on nodejs<->winapi for Windows to have native UI there. I also wanted to do this in the *nix space too, but of course if there are solutions already I will leverage them.

I see there is node-gjs, but what is the difference between the two?

Thanks a bunch @WebReflection.

**Comment by WebReflection** [comment]
if by `node-gjs` you mean [jsgtk](https://www.npmjs.com/package/jsgtk) these two are different worlds.

This repo has been abandoned because I had no idea how to contribute on the C side of it, I could only contribute on the JS one.

That's pretty much the end of the story. It would've been cool, but not enough interest around this, not enough developers onboard, and tools like [Electron](https://electron.atom.io) for many are all they need to create native-like UIs with nodejs horses-power behind the scene.

Of course if anyone would like to help, I'd be happy to collaborate, but so far that didn't happen, many tried to GIR via node and JS, and not a single successful story came out so far.

:man_shrugging:

**Comment by lf94** [comment]
Sorry yes, I meant jsgtk.

The reason why I want to avoid Electron is because the binary size and efficiency is terrible. I've used it for 2 applications and yeah it's been awesome, but like I said has some major downsides.

What do you need done on the C side?

What's the downside of pursuing jsgtk vs this project?

**Comment by WebReflection** [comment]
jsgtk is a "nodefication" of GJS which is based on kinda jurassic syntax, has a very unfriendly module system, and most common, and simple, functionalities provided by nodejs core are absent, **but** it is not based on V8 or NodeJS at all, so it will always lack updates, syntax speaking, performance, and real-world module compatibility, which is what this repository would've solved with ease.

What needs to be done here? Well, @magcius is the only one that can answer that, which is why I'm not good at moving this project forward: I don't even know what to do.

I can do unimaginable things on the JS side though 🎉 😄 🍻

**Comment by lf94** [comment]
Cool, @magcius answer so I can bring this puppy back to life.

**Comment by benwaffle** [comment]
@romgrk has made some progress in [his fork](https://github.com/romgrk/node-gtk), but that is also abandoned

**Comment by magcius** [comment]
I no longer work at Endless or on Linux stuff

**Comment by WebReflection** [comment]
@magcius sad to hear that, I hope it's for better opportunities.

@benwaffle thanks for the link, it seems a repetitive pattern though.

  1. start implementing it
  2. get stuck and abandon

AFAIK [node-gir](https://github.com/creationix/node-gir) had the exact same pattern too.

I wonder if @romgrk can explain what's missing or needed in his fork to move forward, if even possible.

**Comment by lf94** [comment]
I'll begin from his fork I suppose! Thank you everyone.

**Comment by WebReflection** [comment]
if you need any help on the JS side please let me know

**Comment by lf94** [comment]
@WebReflection yeah I will. Anything to move this along faster...

Today I updated the dependencies and made the C++ compile with no compiler errors.

Now I'm running into segmentation fault issues, that seem to be caused by the garbage collector. When I make a global reference to some variables, they go away (even that one V8Array error @romgrk was having). 

Now I'm fighting this:

var dependencies = GIRepository.Repository_get_immediate_dependencies.call(repo, ns, version);
TypeError: Cannot read property 'call' of undefined

**Comment by magcius** [comment]
That means the function `Repository_get_immediate_dependencies` doesn't exist. Something's going wrong when bootstrapping `GIRepository`. Double-check the GIR file, that it has that method in there, and double-check that the code that bootstraps `GIRepository` is correct.

GIRepository is a bit of a wacky special case when it comes to GI libraries: since it's intended to be bootstrapped from it doesn't use GObjects or fancy things much so you'll often see these weird "namespaced" methods. I also do a similar thing during the bootstrap phase. I imagine you can figure it out.

**Comment by lf94** [comment]
Yeah I knew it didn't exist, but didn't know why since it should because it was built before? Was that function deprecated, maybe? To me that conclusion makes no sense since the girepository version is the same...At least, I think it is.

Thank you for pointing me in the right direction @magcius :) There were some posts recently about people being fed-up with Electron apps, so I figured I can get a move on with this, create something presentable (a simple text editor?) and get more people on-board.

**Comment by lf94** [comment]
Check this out: https://github.com/creationix/node-gir !

**Comment by WebReflection** [comment]
We all know it, it's also abandoned

**Comment by romgrk** [comment]
Yo @lf94, here are some missing things on my fork:
 - Handle C-style out parameters (that is when a you pass a pointer to a function as an argument, but it is used as a return value). This needs some thinking to adapt it to javascript.
 - Boxed Unions support: This also needs some thinking, I haven't been able to pin a nice way to handle unions. For the moment, prototype must be set by hand (in JS) to cast an object into a type. E.g. `event.__proto__ = Gdk.EventKey` or `event.__proto__ = Gdk.EventMouse`. (though normal boxeds work fine)
 - GIInterfaces: I'm not sure what those are
 - Virtual functions: Never had time to look into this
 - Make JS play nice with GObject classes, eg:
```js
class MyWindow extends Gtk.Window { /* ... */ }
```
 - Possibly more

Also, GTK documentation being what it is, reading through PyGObject source code has been a great source of inspiration for some parts I didn't understand very well.

**Comment by WebReflection** [comment]
> Make JS play nice with GObject classes, eg:
> class MyWindow extends Gtk.Window { /* ... */ }

this was solved in jsgtk, example:
https://github.com/WebReflection/jsgtk/blob/master/examples/extend.js


> reading through PyGObject source code ...

same here when it comes to PyGTK vs GJS

**Comment by romgrk** [comment]
@WebReflection You know what I always thought would be nice to have, but never had time to do? Generate TypeScript definitions of GIR introspectable libraries. 

Other interesting integration: using Gtk + JSX; there is an article here (https://blogs.gnome.org/desrt/2016/08/17/a-quick-look-at-using-jsx-in-gnome/) exploring the concept. It would be sooo nice to be able to just
```jsx
const win = <Gtk.Window title="Hello, World" width="300" height="150" onDestroy={Gtk.mainQuit}>
            <Gtk.Button label="click me" onClicked={buttonClicked} />
          </Gtk.Window>
```

**Comment by WebReflection** [comment]
I have done runtime scan of GIR libraries through GJS/jsgtk and here too.

example: https://github.com/WebReflection/jsgtk/blob/master/jsgtk_modules/jsgtk/gtk_modules.js

or crazy stuff like: https://github.com/WebReflection/jsgtk/blob/master/utils/gir-to-jsgtk-gi.html

The problem I have with jsgtk is that GIR scan at runtime is quite slow while here it was done differently, passing by all the already known info per class ... then I had to patch those anyway to have `camelCase`

I'm patiently waiting for GJS chaps to catch up with modern JS and see how/when they'll integrate GTK and ES `class` , after that I'll try to figure out if there's some way to reproduce that in node too.

TL;DR to be honest, TypeScript and JSX are the least of my problem ... we don't even have a usable Gtk in node now 😞

**Comment by benwaffle** [comment]
gjs is upgrading their version of mozjs, they recently went to spidermonkey 38. They added a `Promise` implementation while they wait to upgrade further
